Use copy of TT entry for consistent values at a node.

locutus2 opened the pull request:
To avoid that a retrieved TT entry changes at a node during the search use a copy of it instead.

For this a new struct TTData is introduced as storage which is simply an extended TTEntry. This data structure contains an additional pointer to the original TT entry for saving.

Passed STC (non-regression):
LLR: 2.93 (-2.94,2.94) <-1.75,0.25>
Total: 69248 W: 18020 L: 17845 D: 33383
Ptnml(0-2): 176, 7385, 19332, 7550, 181
https://tests.stockfishchess.org/tests/view/665c2c87ab5608f07fdcca6e

Passed LTC (non-regression):a
LLR: 2.95 (-2.94,2.94) <-1.75,0.25>
Total: 219006 W: 55631 L: 55613 D: 107762
Ptnml(0-2): 88, 21934, 65447, 21940, 94
https://tests.stockfishchess.org/tests/view/665c43b304452685e60249bf

Passed SMP STC (non-regression):
LLR: 2.94 (-2.94,2.94) <-1.75,0.25>
Total: 51136 W: 13344 L: 13149 D: 24643
Ptnml(0-2): 45, 5794, 13705, 5969, 55
https://tests.stockfishchess.org/tests/view/665c507704452685e6024ebc

Passed SMP LTC (non-regression):
LLR: 2.94 (-2.94,2.94) <-1.75,0.25>
Total: 157240 W: 40229 L: 40151 D: 76860
Ptnml(0-2): 23, 16171, 46155, 16247, 24
https://tests.stockfishchess.org/tests/view/665c7021fd45fb0f907c214a

Bench: 1369817

Sopel97 commented:
Considering what passed and what didn't, I'm in favor of this approach. Minimal required changes so far, but not out line of the ideal path.

peregrineshahin commented:
This patch extends the same type/class/struct/concept of an object only to fake a copy that is not needed in the first place.
This is how I expect the code to coded if someone is afraid to look at the actual meanings and assumptions of the original code.

locutus2 commented:
@Sopel97 
Perhaps you can retry your version without the copy of TTEntry and update the members directly from the original tt entry. This gives a bigger interval where the  entry could be overwritten but i think not much (for single core it would be the same).

MinetaS commented on src/tt.h:
```diff
@@ -63,6 +63,20 @@ struct TTEntry {
+
+    TTData(TTEntry* tte) :
+        TTEntry(*tte),
+        ttEntry(tte){};
```
I prefer `entry` as a member name over `ttEntry`. Also a semicolon after braces is redundant.

locutus2 commented on src/search.cpp:
```diff
@@ -606,22 +605,23 @@ Value Search::Worker::search(
-    ttCapture = ttMove && pos.capture_stage(ttMove);
+
+    TTData tte = tt.probe(posKey, ss->ttHit);
+    ttValue    = ss->ttHit ? value_from_tt(tte.value(), ss->ply, pos.rule50_count()) : VALUE_NONE;
```
I can try this but then it become similar to Sopels test but for me this approach seems clearer. The additional values are not used in all places like in qsearch and bundles for my feeling to much search related logic there. For me its a simply immutable wrapper for the tt entry which should not have more logic. But i will make a version and test so maintainers have more different versions for decision. The question if this version has then the same problem as Sopel's one to pass.

Disservin replied:
I think we are talking about different things.. I was just asking to about moving
`Move ttMove` from here https://github.com/official-stockfish/Stockfish/blob/e23913a9a5489f766558665a8951fad221e23761/src/search.cpp#L550
to here https://github.com/official-stockfish/Stockfish/blob/e23913a9a5489f766558665a8951fad221e23761/src/search.cpp#L611

MinetaS replied:
or perhaps we give `ttMove`, `ttValue` and `ttCapture` different names? Yes, these variables are used in search only while constructed based on information of a TT entry. With this approach I'd like to leave these outside `TTData`, but if one can wrap all variables into a single struct without losing readability/performance, that's still worth to consider.

Disservin replied:
Mh different names, dunno which ones lol? 
I was just saying all this since we now initialize the tt data directly we might as well just move the other tt related logic down to where it also first assigned. Basically moving away from this legacy C requirement where variables had to be declared at the start of a function.

locutus2 replied:
@Disservin 
That i misunderstand. Yes that can we do. In my C++ code i prefer this too(if possible because of scope). 'Init at declaration'. Its anyway safer by avoiding addicently undefined behavior.

locutus2 replied:
Refactored PR with @Disservin  and @MinetaS  comments.

dubslow commented:
For the record, some other variants:

Sopel: 
https://tests.stockfishchess.org/tests/view/665c707ffd45fb0f907c2167
https://tests.stockfishchess.org/tests/view/665c536b04452685e6024ef1
https://tests.stockfishchess.org/tests/view/665c69553542f91ad1c54fea

peregrine:
https://tests.stockfishchess.org/tests/view/665ce772fd45fb0f907c538e
https://tests.stockfishchess.org/tests/view/665d28a7fd45fb0f907c5495

and now joster too:
https://tests.stockfishchess.org/tests/view/665e0374fd45fb0f907c77e1

Frankly I prefer all of these variants to this PR, I find this PR difficult to understand (that the constructor copy is the key bit).

But more importantly, it's not clear to me why certain Sopel variants failed where this passed (and why pere's is currently failing LTC, tho maybe it will turn around). Ideally we need to understand why some fail and some pass

locutus2 commented:
@dubslow 
But using the copy constructor is exactly the C++ way to do it (at default it simply a verbatim copy of the object (perhaps using memcpy or some specialzed code). So we should do this and not the old C-style. That would be a step backward. But perhaps a comment could help.

dubslow commented:
Frankly, as one of sopel's tests discovered that `ttMove` and `ttValue` have significantly different behavior than `ttDepth` and `ttBound`, I hope that joster's unstructured version works best. We were on the structured path because sopel had suggested it, but some simply behave different than others, so joster's version is best imo

locutus2 commented:
But that solution let the dev use the tte directly again. The advantage of my solution we have the copy and can only use them directly (we can save to the TT to change the values). So in this sense its more robust. And from my experience it helps to avoid errors using more robust code.

ddobbelaere commented on src/tt.h:
```diff
@@ -63,6 +63,20 @@ struct TTEntry {
 
+struct TTData: public TTEntry {
+
+    TTData(TTEntry* tte) :
```
I feel it's safer to mark this single-parameter constructor as `explicit`, so that no implicit conversions from `TTEntry *` are possible by accident.

locutus2 replied:
That is a good idea. Makes the code more robust against errors.

locutus2 commented:
I have done a small change by reverting the probe function back to master version and gives this now directly as parameter for TTData. This has the advantage that we can retrieve the raw TTEntry pointer if needed but at the same time avoids accidental use of the raw pointer in current search (where its not needed and only a potenial source of an error).

peregrineshahin commented:
it looks like it's undoable to copy the whole entry or values at once without failing the thread sanitizer, there is something wrong with all solutions presented except the (mineta/jooster one which will not fail the thread sanitizer)

vondele commented:
please add the needed function to the tsan exceptions https://github.com/official-stockfish/Stockfish/blob/ba06671aa9df5c0a3fa5f1fa2ce17ea4aa742b7a/tests/instrumented.sh#L41-L53

vondele commented:
> I have done a small change by reverting the probe function back to master version and gives this now directly as parameter for TTData. This has the advantage that we can retrieve the raw TTEntry pointer if needed but at the same time avoids accidental use of the raw pointer in current search (where its not needed and only a potenial source of an error).

I just like the pointer was hidden in the previous solution. It really shouldn't be accessible to search.

locutus2 commented:
Interesting after my last change the sanitizer fails:
```
WARNING: ThreadSanitizer: data race (pid=3789)
  Read of size 1 at 0x7fa967309830 by thread T2:
    #0 Stockfish::TTData::TTData(Stockfish::TTEntry*) <null> (stockfish+0x30d5b)
    #1 int Stockfish::Search::Worker::qsearch<(Stockfish::NodeType)0>(Stockfish::Position&, Stockfish::Search::Stack*, int, int, int) /home/runner/work/Stockfish/Stockfish/src/search.cpp:1466 
```
Have to investigate.

vondele commented:
> Have to investigate.

I think it is no surprise, I was surprised that previously it didn't complain when the data was copied in the constructor.

locutus2 commented:
@vondele 
It is hidden because we have currently no variable to access. But the original TTEntry pointer can be saved and then used if needed. Here in search i see no reason for it because it can only generate inconsistencies. But par example in the extract_ponder_move routine i reverted back to the direct probe because only the tt move is retrieved.

But the new version shows now some sanitizer problem. I have updated with master but it fails agains. So i will go back to my previous version.

locutus2 commented:
@vondele 
Thanks, then my original version had no problem i have done the copy in the probe function which has already an exception. But i had already reverted back to the previous version. We can then discuss which seems better.

753339778 reviewed (changes requested):
B

vondele commented:
closing in favor or  #5369
